block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -267,144 +347,137 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: BlockStore] private
 
   private def updateDataLookupFile(blockMetadata: BlockMetadata): F[Unit] =
     for {
-      dataLookupCrc          <- blockMetadataCrcRef.get
+      dataLookupCrc          <- getBlockMetadataCrc
       blockBytes             = blockMetadata.toByteString
       toAppend               = blockBytes.size.toByteString.concat(blockBytes).toByteArray
-      dataLookupOutputStream <- blockMetadataLogOutputStreamRef.get
-      _                      <- Sync[F].delay { dataLookupOutputStream.write(toAppend) }
-      _                      <- Sync[F].delay { dataLookupOutputStream.flush() }
+      dataLookupOutputStream <- getBlockMetadataLogOutputStream
+      _                      <- dataLookupOutputStream.write(toAppend)
+      _                      <- dataLookupOutputStream.flush
       _                      <- dataLookupCrc.update(toAppend)
       _                      <- updateDataLookupCrcFile(dataLookupCrc)
     } yield ()
 
   private def updateDataLookupCrcFile(newCrc: Crc32[F]): F[Unit] =
     for {
       newCrcBytes <- newCrc.bytes
-      _ <- Sync[F].delay {
-            val tmpCrc = Files.createTempFile("rchain-block-dag-file-storage-data-lookup-", "-crc")
-            Files.write(tmpCrc, newCrcBytes)
-            Files.move(tmpCrc, blockMetadataCrcPath, StandardCopyOption.REPLACE_EXISTING)
-          }
+      tmpCrc      <- createTemporaryFile[F]("rchain-block-dag-file-storage-data-lookup-", "-crc")
+      _           <- writeToFile[F](tmpCrc, newCrcBytes)
+      _           <- replaceFile(tmpCrc, blockMetadataCrcPath)
     } yield ()
 
-  def getRepresentation: F[BlockDagRepresentation[F]] =
+  private def representation: F[BlockDagRepresentation[F]] =
     for {
-      _              <- lock.acquire
-      latestMessages <- latestMessagesRef.get
-      childMap       <- childMapRef.get
-      dataLookup     <- dataLookupRef.get
-      topoSort       <- topoSortRef.get
-      sortOffset     <- sortOffsetRef.get
-      _              <- lock.release
+      latestMessages <- getLatestMessages
+      childMap       <- getChildMap
+      dataLookup     <- getDataLookup
+      topoSort       <- getTopoSort
+      sortOffset     <- getSortOffset
     } yield FileDagRepresentation(latestMessages, childMap, dataLookup, topoSort, sortOffset)
 
+  def getRepresentation: F[BlockDagRepresentation[F]] =
+    lock.withPermit(representation)
```

Comment:
Is it necessary to obtain a lock if we are doing only read-only operations here?

Reply:
Is it because each individual item could be updated and so we might get some old state mixed with new state?

Reply:
Yes, you could indeed end up grabbing a mixed state. It is possible to fix this since we have removed  all`Ref`s and replaced them with a single `MonadState`. So you could get a consistent state from `MonadState` and get all individual items from it. However, this change is not as simple as just rewriting `representation` method because all other state-modifying methods assume that they "own" this state during execution and partially update it like this:

```scala
...
  _ <- setDataLookup(Map.empty)
  _ <- setChildMap(Map.empty)
  _ <- setTopoSort(Vector.empty)
  _ <- setLatestMessages(Map.empty)
...
```

So we have to rewrite all these places to modify `MonadState` atomically. Which might actually make sense but probably deserves its own PR.